### PR TITLE
Enhancing Topic Metadata Accessibility in the Hedera REST API

### PR DESCRIPTION
**Pull request by kantorcodes**
**Description**:
Introduces a entity lookup route for individual TopicIds along with basic information about those topics. By incorporating this change dApps will be able to tie in enhanced functionality without needing to operate a Mirror Node.

### CODE REVIEW

File: HIP/hip-874.md [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+}
+```
+
+The addition of memo, created_timestamp, last_updated_time,  deleted, admin_key , auto_renew_period, submit_key ,auto_renew_account_id,id fields will provide comprehensive metadata about topics directly through the API.
+
+### Rationale:
+Providing direct access to topic ID metadata, including the memo and other relevant information, addresses a critical need for developers building on the Hedera network. This enhancement will facilitate a range of applications, from simple messaging systems to complex, multi-layered applications requiring detailed audit trails. The proposal supports the principles of transparency, efficiency, and developer friendliness that are essential to the Hedera ecosystem.
+
+### Backwards Compatibility:
+This change is designed to be backwards compatible, adding new fields to the API response without altering existing fields or endpoint behaviors. This ensures that existing applications will continue to operate without modification, while new and updated applications can take advantage of the enhanced metadata accessibility.
```
Comment (steven-sheehy) [inline HIP/hip-874.md:50]: It doesn't add new fields to the API response. It adds a whole new API, which is of course backwards compatible.

File: HIP/hip-874.md [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+{"_status":{"messages":[{"message":"Not found"}]}}
+```
+
+Proposed Response Structure:
+```json
+{
+  "id": "0.0.123456",
+  "memo": "[Topic Memo]",
+  "admin_key": null,
+  "deleted": "[Boolean]",
```
Comment (steven-sheehy) [inline HIP/hip-874.md:36]: It's not a string
```suggestion
  "deleted": false,
```

File: HIP/hip-874.md [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+  "submit_key": null,
+  "auto_renew_period": 7890000,
+  "auto_renew_account_id": "0.0.654321",
```
Comment (steven-sheehy) [inline HIP/hip-874.md:39]: To match similar field in other APIs:

```suggestion
  "auto_renew_account": "0.0.654321",
```

File: HIP/hip-874.md [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+```
+
+Proposed Response Structure:
+```json
+{
+  "id": "0.0.123456",
+  "memo": "[Topic Memo]",
+  "admin_key": null,
+  "deleted": "[Boolean]",
+  "submit_key": null,
```
Comment (steven-sheehy) [inline HIP/hip-874.md:37]: Show an example so it's clear the format. Same for admin_key
```suggestion
  "submit_key": {
      "_type": "ED25519",
      "key": "bfc9d4a344389324119eed485eb3a875b1e5a54bd189d2fd5ad88ce1d3473dee"
  },
```

File: HIP/hip-874.md [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+
+Current Response Structure returns:
+```json
+{"_status":{"messages":[{"message":"Not found"}]}}
+```
+
+Proposed Response Structure:
+```json
+{
+  "id": "0.0.123456",
```
Comment (steven-sheehy) [inline HIP/hip-874.md:33]: To match other APIs:
```suggestion
  "topic_id": "0.0.123456",
```

File: HIP/hip-874.md [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+  "memo": "[Topic Memo]",
+  "admin_key": null,
+  "deleted": "[Boolean]",
+  "submit_key": null,
+  "auto_renew_period": 7890000,
+  "auto_renew_account_id": "0.0.654321",
+  "created_timestamp": "[Creation Time]"
```
Comment (steven-sheehy) [inline HIP/hip-874.md:40]: Show an example
```suggestion
  "created_timestamp": "1707953087.386834512"
```

**Review by steven-sheehy — CHANGES_REQUESTED** [review]

File: HIP/hip-874.md (hunk 1 above)
Comment (kantorcodes) [reply HIP/hip-874.md]: Yup fair enough haha. Updated the language here.

**Comment by mgarbs** [comment]
This has some merge conflicts @kantorcodes

File: HIP/hip-874.md [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+```json
+{"_status":{"messages":[{"message":"Not found"}]}}
+```
+
+Proposed Response Structure:
+```json
+{
+  "topic_id": "0.0.123456",
+  "memo": "[Topic Memo]",
+  "admin_key": null,
```
Comment (steven-sheehy) [inline HIP/hip-874.md:35]: Also show an example for an admin_key

File: HIP/hip-874.md [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+      "_type": "ED25519",
+      "key": "bfc9d4a344389324119eed485eb3a875b1e5a54bd189d2fd5ad88ce1d3473dee"
+  },
+  "auto_renew_period": 7890000,
+  "auto_renew_account": "0.0.654321",
+  "created_timestamp": "1707953087.386834512"
+}
+```
+
+The addition of memo, created_timestamp, last_updated_time,  deleted, admin_key , auto_renew_period, submit_key ,auto_renew_account_id,id fields will provide comprehensive metadata about topics directly through the API.
```
Comment (steven-sheehy) [inline HIP/hip-874.md:47]: Some field names no longer match. Either remove explicit field name list or adjust.

File: HIP/hip-874.md [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+
+## Motivation:
+The ability to access topic ID metadata, including the memo field, submit key, and deleted field directly through the REST API is crucial for a wide range of applications, including messaging, event tracking, and audit-ability functions. Currently, developers may need to employ indirect methods or multiple API calls to retrieve this information, leading to inefficiencies and potential inaccuracies. By providing direct access to this metadata, the proposal seeks to enhance developer experience, reduce computational load, and foster the development of more feature-rich applications on the Hedera platform.
+
+## Specification:
+The proposed enhancement involves extending the REST API to include a new endpoint or modify existing endpoints to return topic ID metadata, including the memo field and other relevant entity information.
```
Comment (steven-sheehy) [inline HIP/hip-874.md:20]:
```suggestion
The proposed enhancement involves extending the REST API to include a new endpoint to return topic ID metadata, including the memo field and other relevant entity information.
```

**Review by steven-sheehy — CHANGES_REQUESTED** [review]

File: HIP/hip-874.md [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+      "_type": "ED25519",
+      "key": "bfc9d4a344389324119eed485eb3a875b1e5a54bd189d2fd5ad88ce1d3473dee"
+  },
+  "deleted": false,
+  "submit_key": {
+      "_type": "ED25519",
+      "key": "bfc9d4a344389324119eed485eb3a875b1e5a54bd189d2fd5ad88ce1d3473dee"
+  },
+  "auto_renew_period": 7890000,
+  "auto_renew_account_id": "0.0.654321",
```
Comment (steven-sheehy) [inline HIP/hip-874.md:45]: We seem to be going in review circles...

```suggestion
  "auto_renew_account": "0.0.654321",
```

File: HIP/hip-874.md [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+  "updated_timestamp" : "1707953087.386834512",
```
Comment (steven-sheehy) [inline HIP/hip-874.md:46]: Our standard is to use a timestamp range. This leaves open the possibility to support a `timestamp` query parameter for historical views of the topic.

```suggestion
  "timestamp" : {
    "from: "1707953087.386834512",
    "to": null
  }
```

File: HIP/hip-874.md [hunk 12]
```diff
@@ -0,0 +1,60 @@
+---
+hip: 874
+title: Enhancing Topic Metadata Accessibility in the Hedera REST API
+author: Michael Kantor (@kantorcodes), Ty Smith (@ty-swirldslabs)
+type: Standards Track
+category: Mirror
+status: Draft
+created: 2024-02-14
+requested-by: TierBot
+discussions-to: https://github.com/hashgraph/hedera-improvement-proposal/pull/883
```
Comment (steven-sheehy) [inline HIP/hip-874.md:10]: Add a `needs-council-approval: Yes`

**Review by steven-sheehy — CHANGES_REQUESTED** [review]

**Review by steven-sheehy — APPROVED** [review]
LGTM

**Pull request merged**